engine/src/main/java/org/hibernate/search/engine/mapper/mapping/building/impl/IndexSchemaFilter.java

```diff
@@ -203,19 +204,60 @@ else if ( parent != null ) {
 			}
 			else {
 				String path = parent.getPathFromSameIndexedEmbeddedSinceNoCompositionLimits( definition );
-				return path == null ? null : path + this.definition.relativePrefix();
+				if ( path == null ) {
+					return null;
+				}
+
+				String pathToCheck = pathLeadingToCurrentNode();
+				if ( isPotentiallyExcludedPath( pathToCheck ) ) {
+					return null;
+				}
```

Comment:
I wonder why we would check an absolute path (`pathLeadingToCurrentNode()`) against the (relative) exclude paths... I don't think this makes sense?
I think you should just pass `definition.relativePrefix()` to `isPotentiallyExcludedPath()`, and inside that method prepend `this.relativePrefix` to the path every time you recurse into a parent.

If you do that, you will probably be able to move this check to the `if` branch just above? It's recursive, so I would prefer not executing it in a block that is itself executed recursively... That'll give you `O(n²)` performance, not great.

Actually if you do that you probably *have* to move this check to the `if` branch just above. Otherwise I think you'll fail to detect a cycle in some models... I pushed additional tests to my branch.

Reply:
> I wonder why we would check an absolute path (pathLeadingToCurrentNode()) against the (relative) exclude paths... I don't think this makes sense?

I was thinking about it like this -- we should stop going through parents as soon as one of them has an exclude filter that may break a cycle eventually. Then I suppose something like this:
```
class A {
    @IndexedEmbedded(excludePaths="a.a.a.a.a.a.a.ahaha;  // :) 
    A a;
    @Keyword
    String ahaha;
}
```
won't stop the recursion, if we'd be only checking the relative path -- we'd go to a parent, and the exclude starts with `a.` so we are happy and we let it continue nesting more `A a`...
So I thought if we'd do the absolute path -- eventually appending `a.` will go beyond what is in that exclude path and we'd say "No no no -- doesn't look like this cycle will ever break let's fail"

Reply:
I see where you're coming from, but I'm afraid comparing relative paths with absolute paths like that is an approximation that could lead to more bugs... because that's not quite "correct".

By the way, please add a test for your example if you didn't already!

Okay, other idea: when we detect there' are no composition limits, we look for the **furthest** (not closest) occurrence of the same indexed-embedded definition, and we check whether that definition or any ancestor has any excludePaths that is still a prefix for the current path. If there is an infinite recursion, we will detect it eventually. WDYT?

Reply:
Yeah, I've been thinking towards something similar ( finding the longest cycle (even if repeating) so far and then doing some similar checks to what's in there right now using that cycle path without the first element, starting from the current node and going through all the parents). I think there might be a problem if  we only check the furthest same-type-node and its ancestors, we might not catch a case like:
```
class A {
    @IndexedEmbedded
    B b; // (1)
}
class B {
    @IndexedEmbedded
    C c;   
}
class C {
    @IndexedEmbedded(excludePaths= "c.b")
     B b; // (2)
}
```
when we are at `A.b.c.b`  (2), we'll find that b at (1) is the furthest but b at (1) doesn't break the cycle and there's nothing before it ...

So for now I'm trying to come up with some other potential cycle problems to see if the idea breaks 😄

Reply:
> when we are at A.b.c.b (2), we'll find that b at (1) is the furthest but b at (1) doesn't break the cycle and there's nothing before it ...

We will not, since those are different indexed-embedded definitions. See `IndexedEmbeddedDefinition.equals`, it takes the defining type (and relative prefix) into account.